### PR TITLE
Modification to allow deserialization of derrived types

### LINKED ISSUE #11083: Seeking clarification on documentation around System.Xml.Serialization
This is not really an issue, I am posting here due to the line on the page https://www.mono-project.com/community/ which says 

> Feel free to file an issue on GitHub if you want to discuss something, even if it's not a bug.

### Current Behavior

Currently, I have a PCL with a System.Xml.Serialization XmlSerializer constructed with the XmlSerializer(Type, XmlAttributeOverrides) constructor so I can modify how a list of subclasses gets serialized/deserialized.  Essentially, each of my classes derives from the same abstract class, and each of them have a List property (called "children") of the abstract class type.   My XmlAttributeOverrides allows the XmlSerializer to deserialize XML in which any node can be within any other and the enclosing node will have its "children" property deserialized to contain any enclosed nodes. 

This works as I expect on my windows project which references the PCL but my android project which references the same PCL does not work.

I noticed the docs (https://www.mono-project.com/docs/tools+libraries/libraries/xml/) state that:

> Because of lack of ambiguity detection, there is no way to provide XmlMapping XmlTypeMapping/XmlMemberMapping). But If anyone is interested in such effort, integration with XmlSerializer would be interesting task.

Does that apply to my situation?

I basically asked the same thing here: https://stackoverflow.com/questions/52748779/mono-android-xmlserializer-does-not-use-xmlattributeoverrides

## On which platforms did you notice this

[ ] macOS
[ X ] Linux
[ ] Windows

**Version Used**:

I am using Xamarin.Android 8.2.0.16

This may be another source of my confusion as I am uncertain about the difference between Mono and Xamarin and where each is used/for what each is responsible.

For instance, would the same src code in mcs/class/System.XML be referenced, for instance, in both a Xamarin.Android AND Xamarin.iOS project.

I'm sorry if this issue is inappropriate or if I am even more confused than I think I am :)

### PR DESCRIPTION
**Pull request by majellin24**
Hope I did this correctly. Sorry if not.

Fixes #11083

### CODE REVIEW

File: mcs/class/System.XML/System.Xml.Serialization/XmlTypeMapping.cs [hunk 1]
```diff
@@ -413,11 +424,15 @@ public XmlTypeMapElementInfo GetElement(string name, string ns)
...
-			return null;
+            foreach (XmlTypeMapElementInfo info in _elements.Values)
+                if (info.ElementName == name && info.Namespace == ns)
+                    return info;
+                else if (info.MappedType.DerivedTypes.Count > 0) {
+                    foreach (XmlTypeMapping derrivedInfo in info.MappedType.DerivedTypes)
+                        if (derrivedInfo.ElementName == name && derrivedInfo.Namespace == ns)
+                            return info;
+                }
+            return null;
```
[the remark this answers is not in the document]
Comment (majellin24) [reply mcs/class/System.XML/System.Xml.Serialization/XmlTypeMapping.cs]: I'm not sure if that worked or how to see, sorry!

[the remark this answers is not in the document]
Comment (majellin24) [reply mcs/class/System.XML/System.Xml.Serialization/XmlTypeMapping.cs]: Sorry about that!

**Comment by alexanderkyte** [comment]
I don't think you can see them directly through Github (correct me if I'm wrong), so I'll let you know how the CI results come back.

**Comment by alexanderkyte** [comment]
@monojenkins approve

**Comment by alexanderkyte** [comment]
@monojenkins build

**Comment by majellin24** [comment]
Guessing the null ref is line 430

**Comment by alexanderkyte** [comment]
So mostly green, seeing two tests looking a bit unstable. I'm going to rerun them before telling you there's definitely a breakage. It doesn't look related.

**Comment by alexanderkyte** [comment]
And we often like to squash our commits with a force push, but I believe we have some automation for that once it's green. We can do that on our side if you want.

**Comment by alexanderkyte** [comment]
@monojenkins build failed

**Comment by alexanderkyte** [comment]
It looks like https://github.com/mono/mono/blob/master/mono/tests/bug-2907.cs is failing on FullAOT. 

I'm not familiar enough here with the XML namespace to know if we should blame the test or the PR. It seems to be fine when using the JIT / running normally. 

@kg seems to have done work here recently, they may be a good resource

**Comment by majellin24** [comment]
I made another mistake :(

**Comment by kg** [comment]
There's weird whitespace damage here that makes it hard to read the diffs. I think you converted tabs to spaces? You should undo that once you get a chance, so that the diffs are only lines you actually changed. (If individual lines just had blatantly bad indentation - it looks like one of them did - then it seems fine to fix that one.)

**Comment by majellin24** [comment]
Yeah, I apologize for that mess!  Visual studio made it look like the indentation was correct while I was working but any lines it auto-formatted got 4 space tabs rather than 8, but it didnt look any different until I pushed it.  Then it took me a few commits to understand what was going on and my git skills are not great.  Sorry!!

I will try to make it right when I have a little more time.

In the meantime, I think I fixed the failures from the last round of tests but I have no idea what these new ones are, I dont think it failed on those last time.

**Comment by luhenry** [comment]
@marek-safar can somebody on your team please take a look.

**Comment by kg** [comment]
This looks okay to me but I am not an XML Serializer expert. The two new for-loops added look like they could be a performance issue for people processing lots of XML, but that might be okay if this is a new capability and not (i think?) a risk of regressing performance for existing use cases.

**Comment by luhenry** [comment]
@kg so approve or request changes?

**Comment by kg** [comment]
If nobody else knows XML Serializer better than me, I'd prefer that @majellin24 add at least one automated test for this change. At the very least that will ensure we don't break the feature later.

**Comment by marek-safar** [comment]
@majellin24 could you please add a test for this case for example to https://github.com/mono/mono/blob/master/mcs/class/System.XML/Test/System.Xml.Serialization/DeserializeTests.cs

**Comment by marek-safar** [comment]
Thank you for the contribution

**Pull request merged**